sdk/src/job_runner.rs

```diff
@@ -0,0 +1,145 @@
+use std::pin::Pin;
+
+pub struct MultiJobRunner {
+    pub(crate) enqueued_job_runners: EnqueuedJobRunners,
+    pub(crate) test_mode: bool,
+}
+
+pub type EnqueuedJobRunners = Vec<
+    Pin<
+        Box<
+            dyn SendFuture<
+                Output = Option<tokio::sync::oneshot::Receiver<Result<(), crate::Error>>>,
+            >,
+        >,
+    >,
+>;
+
+pub trait SendFuture: Send + Future + 'static {}
+impl<T: Send + Future + 'static> SendFuture for T {}
+
+pub struct JobBuilder<'a, T> {
+    pub(crate) register_call: Option<RegisterCall>,
+    context: Option<T>,
+    runner: &'a mut MultiJobRunner,
+}
+
+pub(crate) type RegisterCall = Pin<Box<dyn SendFuture<Output = Result<(), crate::Error>>>>;
+
+impl<T> JobBuilder<'_, T> {
+    pub fn with_registration<Fut: SendFuture<Output = Result<(), crate::Error>>>(
+        &mut self,
+        register_call: fn(T) -> Fut,
+    ) -> &mut Self {
+        let context = self
+            .context
+            .take()
+            .expect("Cannot call this function twice");
+        let future = register_call(context);
+        self.register_call = Some(Box::pin(future));
+        self
+    }
+
+    pub fn add_job<K: InitializableEventHandler + Send + 'static>(
+        &mut self,
+        job_runner: K,
+    ) -> &mut MultiJobRunner {
+        let registration = self.register_call.take();
+        let test_mode = self.runner.test_mode;
+
+        let task = Box::pin(async move {
+            if let Some(registration) = registration {
+                // Skip registration if in test mode
+                if !test_mode {
+                    if let Err(err) = registration.await {
+                        crate::error!("Failed to register job: {err:?}");
+                        return None;
+                    }
+                }
+            }
+
+            job_runner.init_event_handler().await
+        });
+        self.runner.enqueued_job_runners.push(task);
+
+        self.runner
+    }
+}
+
+impl MultiJobRunner {
+    pub fn new(env: &GadgetConfiguration<parking_lot::RawRwLock>) -> Self {
+        Self {
+            enqueued_job_runners: Vec::new(),
+            test_mode: env.test_mode,
+        }
+    }
+
+    pub fn with_job_context<'a, 'b: 'a, T: 'b>(&'a mut self, context: T) -> JobBuilder<'a, T> {
+        JobBuilder {
+            register_call: None,
+            runner: self,
+            context: Some(context),
+        }
+    }
+
+    /// Adds a job to the job runner
+    ///
+    /// let x_square = blueprint::XsquareEventHandler {
+    ///      service_id: self.env.service_id.unwrap(),
+    ///      client: client.clone(),
+    ///      signer: signer.clone(),
+    /// };
+    ///
+    /// let x_square2 = blueprint::XsquareEventHandler {
+    ///      service_id: self.env.service_id.unwrap(),
+    ///      client: client.clone(),
+    ///      signer,
+    /// };
+    ///
+    /// MultiJobRunner::new().with_job().add_job(x_square).with_job().add_job(x_square2).run().await;
```

Comment:
Is that the current usage? If so, it's a little confusing. It'd probably be better to just pass in your own `JobBuilder` than have `with_job()` create one for you.

```rust
let runner = MultiJobRunner::new()
    .job(JobBuilder::new().job(x_square))
    .job(JobBuilder::new().registration(/* Something */).job(x_square2));
```